### ISSUE TITLE
Calling LockMutex() in a function(callback) called from C

### ISSUE THREAD

**Issue by GWRon** [bmx-ng/bmx-ng#70]
While I played with threading and callbacks in C I've done this:

```C
#include <unistd.h>
#include <stdio.h>


//forward declaration
int __startThread();

//global vars
int threadID;
int updateStreamsEnabled = -1; //-1 = auto activate on registering a callback
int updateInterval = 100;

//definition and variable for callback
typedef int (*intCallback)(void);
intCallback updateStreamManagerCallback;




int stopThread() {
	updateStreamsEnabled = 0;
	return 1;
}


int startThread(){
	updateStreamsEnabled = 1;
	return __startThread();
};


int RegisterUpdateStreamManagerCallback ( intCallback callback ) {
	updateStreamManagerCallback = callback;
	return 1;
}

void *updateStreamManager(void *v) {
	while(updateStreamsEnabled) {
		//call the blitzmax function which updates the streams buffers
		if(updateStreamManagerCallback != NULL)
			updateStreamManagerCallback();

		//wait some milliseconds till next update
		usleep(updateInterval * 1000);
	}
	return NULL;
}

#ifdef __linux
	#include <pthread.h>

	pthread_t  thread;
	
	int __startThread(){
		fprintf( stderr, "starting thread" );
		threadID = pthread_create(&thread, NULL, updateStreamManager, NULL);	
		return 0;
	};
#endif

#ifdef _WIN32
	#include <windows.h>

	HANDLE  thread;
	
	int __startThread(){
		thread = CreateThread(0, 0, updateStreamManager, NULL, 0, &threadID);
		return 0;
	};
#endif
```

and within BlitzMax I did:
```BlitzMax
[...]
Extern "C"
'	Function RegisterUpdateStreamManagerCallback:int( cbFunc:int())
	Function RegisterUpdateStreamManagerCallback:int( cbFunc:Byte Ptr )
	Function StartUpdateStreamManagerThread:int() = "startThread"
	Function StopUpdateStreamManagerThread:int() = "stopThread"
End Extern

'...registered the following function as callback:
	Function UpdateStreamManagerCallback:int()
		if not instance then return False

		
		LockMutex(refillBufferMutex)
			instance.RefillBuffers()
		UnlockMutex(refillBufferMutex)

		return True
	End Function
```
This then segfaulted. Removing the `LockMutex` and `UnlockMutex` commands it worked.

How could I use these commands within that function (or within a function called from there) ? Did I do something wrong or am I missing to run an additional command?

**Comment by GWRon** [comment]
Should I wrap the function in a check if it is running (simple integer variable) ... as integer access does not need a mutex?

**Comment by GWRon** [comment]
Ok ... no further checks but it seems to run now.

How to reuse the blitz.mod-definitions for threads? For now I manually added win32/linux support but of course it would be cool to automatically work for Android/ios ... and whatever @woollybah added to blitz.mod/blitz_threads.c/.h (and other files)?

And how to do "differently" for vanilla (if needed)?

**Comment by HurryStarfish** [comment]
You might need to call `BRL.Blitz.GCRegisterMyThread`/`bbGCRegisterMyThread` to let the GC know about the thread you're creating. I don't know if that's what's causing the crash, though.

**Comment by GWRon** [comment]
Yes I thought about that too (it's an NG only feature).

@ LockMutex
The issue is ... that I used `mutex:TMutex = new TMutex` - and in debug it throw an assert - seems one needs to execute `CreateMutex()`.

**Comment by HurryStarfish** [comment]
Yes, the use of `Create` methods/functions was a common workaround for the lack of parameterized constructors in legacy BlitzMax. It's quite ugly and no longer needed in NG - not to mention that it wouldn't have been necessary for mutexes in the first place, since `TMutex.Create` takes no parameters...
We should probably have a standard library cleanup at some point and add proper constructors for everything.

**Comment by GWRon** [comment]
Doing a "cleanup" is a good idea. Only disadvantage is the higher complexity for backports to vanilla - but maybe it is time step out of the shadow of legacy BlitzMax and let it outdate more and more.

**Comment by woollybah** [comment]
The issue with CreateMutex() is that it may return Null if it can't get a handle on one.
If the functionality was in a New(), you can't return Null from there, in which case you would either need to check the returned instance for validity, or Throw from the New.

Perhaps the current documentation/examples are not sufficient?

**Comment by GWRon** [comment]
Yeah...so eg. what to do if the Mutex creation fails already ;-)

The reason for Create() is of course valid now. Thanks for the hint.


Maybe there is a way to "private" the New() for certain types? Or to throw if called from outside (to hint about above's reason).

**Comment by HurryStarfish** [comment]
> If the functionality was in a New(), you can't return Null from there, in which case you would either need to check the returned instance for validity, or Throw from the New.

In that case, it should throw. I don't see a problem with that. Mutex creation isn't normally supposed to fail; if it does, there's a problem, in which case a `Throw` is appropriate. (Preferrably with a proper, unambiguous exception, not just a string like some parts of the standard library tend to do.)

> Maybe there is a way to "private" the New() for certain types?

Sure, you can always do
```
Private
Method New() End Method
```
to "remove" the default constructor of a type. That's unnecessary for types like `TMutex` though; I don't see a problem with giving such types a usable constructor (as the clean solution) in addition to the existing `Create` (for backwards compatibility).
If `New` can throw, that should be documented, of course. The same goes for a `Create` method potentially returning `Null`...
